### ISSUE TITLE
Finding types of expressions yields "unknown" for generic types

### ISSUE THREAD

**Issue by sdroege** [rust-analyzer/rust-analyzer#4189]
```rust
fn main() {
    let a = Some(3);
    let _b = a.and_then(|a1| a.map(|a2| a1 + a2));
    let c = std::sync::Mutex::new(2);
    let d = c.lock().unwrap();
    assert_eq!(*d, 2);
}
```

All variables are detected as "unknown" type here. This also has the side-effect that no autocomplete suggestions are provided (for example when typing `a.and_` it gave nothing, or `c.loc` etc.).

Even when doing `let c: std::sync::Mutex<i32> = ...` the type of `c` is unknown (or the equivalent for `a` with `Option<i32>`.

**Comment by lnicola** [comment]
Hmm..

![image](https://user-images.githubusercontent.com/308347/80486233-81c28d00-8963-11ea-8034-b4d67dfedcb8.png)

**Comment by bjorn3** [comment]
Is the `rust-src` rustup component installed?

```
$ rustup component list | grep rust-src
rust-src (installed)
```

**Comment by sdroege** [comment]
Indeed, that was it. I'm sure I installed it before but apparently not (or not in this machine?). Sorry for the noise!

**Comment by matklad** [comment]
Yet another thing which we can show in status... cc #4074
